float64 x
float64 y
bool safe_flag
bool stop_flag
